Assignment5.java

```diff
@@ -0,0 +1,198 @@
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Queue;
+import java.util.LinkedList;
+import java.util.LinkedHashSet;
+import java.util.Collections;
+
+public class Assignment5 {
+    /**
+     * Infers the alphabet of an unknown language from list of words of that language.
+     *
+     * @param words list of words in lexicographic order.
+     * @return characters in lexicographic order.
+     */
+    public static List<Character> inferAlphabet(List<String> words) {
+        if (words.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        Set<Character> set = new HashSet<>();
+
+        String s1 = words.get(0);
+        // Performed so that all characters will be in the returned list even if topologicalSort doesn't contain some.
+        addAllCharacters(s1, set);
+
+        Map<Character, Vertex> graph = new HashMap<>();
+
+        for (int i = 0; i < words.size() - 1; i++) {
+            s1 = words.get(i);
+            String s2 = words.get(i + 1);
+
+            addAllCharacters(s2, set);
+
+            try {
+                int index = findFirstDifferingIndex(s1, s2);
+                // s1.charAt(index) comes before s2.charAt(index).
+                createEdgeBetweenVertices(s1.charAt(index), s2.charAt(index), graph);
+            } catch (NoDifferenceException ignored) {
+                // Just continue if two strings are identical.
+            }
+        }
+
+        Set<Character> alphabet = topologicalSort(graph);
+
+        alphabet.addAll(set);
+
+        return new ArrayList<>(alphabet);
+    }
+
+    /**
+     * Adds all characters of s to set
+     *
+     * @param s   word whose characters are going to be added to set
+     * @param set holds the characters of s
+     */
+    private static void addAllCharacters(String s, Set<Character> set) {
+        for (int i = 0; i < s.length(); i++) {
+            set.add(s.charAt(i));
+        }
+    }
+
+    private static class NoDifferenceException extends Exception {
+
+    }
+
+    /**
+     * Finds the index at which s1 and s2 start to differ.
+     *
+     * @param s1 string number one.
+     * @param s2 string number two.
+     * @return index s1 and s2 start to differ.
+     * @throws NoDifferenceException if s1 and s2 are not different in an index of the shorter one.
+     */
+    private static int findFirstDifferingIndex(String s1, String s2) throws NoDifferenceException {
+        int bound = Math.min(s1.length(), s2.length());
+        int index = 0;
+
+        while (index < bound && s1.charAt(index) == s2.charAt(index)) {
+            index++;
+        }
+
+        if (index == bound) {
+            throw new NoDifferenceException();
```

Comment:
I feel like using exception does not really fit here - you define special class to signal it, but then ignore it if exception is thrown. Exceptions should be used to handle things that are, well, exceptional (like errors), while it looks like you couldn't decide here whether identical strings are a problem (throw exception) or not (ignore it).

If two identical strings in dictionary are OK, I think it would be better to either use for example `Optional<Integer>` as return value or use `-1` as special value indicating "no differences".

If two identical strings are not OK, just throwing `IllegalArgumentException` and letting program die would be acceptable.